practice_2/WeatherApp/VIew/WeatherViewController.swift

```diff
@@ -0,0 +1,146 @@
+        let location = getLocation(from: gestureRecognizer)
+        let tempUnits = getTemperatureUnits()
+        presenter?.fetchWeather(by: location, units: tempUnits.urlUnits)
+        presenter?.fetchForecast(by: location, units: tempUnits.urlUnits)
+        let temperature = presenter?.weather?.temperature
+        addAnnotation(by: location, temperature: temperature, unit: tempUnits.unit)
+    }
+
+    private func getTemperatureUnits() -> (unit: String, urlUnits: String) {
+        if let unit = UserDefaults.standard.string(forKey: "temperutureUnit") {
+            if unit == "°C" {
+                return (unit: "°C", urlUnits: "metric")
+            } else {
+                return (unit: "°F", urlUnits: "imperial")
+            }
+        } else {
+            return defaultUnits
+        }
+    }
+
+    private func getLocation(from gestureRecognizer: UIGestureRecognizer) -> CLLocationCoordinate2D {
+        let touchLocation = gestureRecognizer.location(in: mapView)
+        let location = mapView.convert(touchLocation, toCoordinateFrom: mapView)
+        return location
+    }
+
+    private func addAnnotation(by location: CLLocationCoordinate2D, temperature: Double?, unit: String) {
+        let annotation = MKPointAnnotation()
+        if let temperature = temperature {
+            annotation.title = "\(temperature) \(unit)"
+        } else {
+            annotation.title = "\(temperature as Double?)"
+        }
+        annotation.coordinate = CLLocationCoordinate2D(latitude: location.latitude, longitude: location.longitude)
+        mapView.addAnnotation(annotation)
+    }
+}
+
+// MARK: MKMapView extension
+extension MKMapView {
+    func centerToLocation(latitude: Double, longitude: Double, regionRadius: CLLocationDistance = 100000) {
+        let location = CLLocation(latitude: latitudeNN, longitude: longitudeNN)
+        let coordinateRegion = MKCoordinateRegion(
+            center: location.coordinate, latitudinalMeters: regionRadius, longitudinalMeters: regionRadius)
+        setRegion(coordinateRegion, animated: true)
+    }
+}
+
+// MARK: MKMapViewDelegate
+extension WeatherViewController: MKMapViewDelegate {
+    func mapView(_ mapView: MKMapView, didSelect view: MKAnnotationView) {
+        let storyboard = UIStoryboard(name: "WeatherApp", bundle: nil)
+        guard let detailsWeatherVC = storyboard.instantiateViewController(
+                identifier: "DetailsWeatherVC") as? DetailsWeatherViewController else { return }
+        detailsWeatherVC.weather = presenter?.weather
+        if let forecastDays = presenter?.forecastDays {
+            detailsWeatherVC.forecastDays = forecastDays
+        }
+        self.navigationController?.pushViewController(detailsWeatherVC, animated: true)
+    }
+}
+
+// MARK: WeatherViewProtocol
+extension WeatherViewController: WeatherViewProtocol {
+   func failure(error: Error) {
+        print("FAIL!!!")
+        print(error.localizedDescription)
```

Comment:
may be it's possible to show some error message to the user in ui,
it's not mandatory, just try to think about possible solution

Reply:
Fixed. Add Alert message.
Thanks.